File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException{
        
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
      BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

      int N = Integer.parseInt(br.readLine());

      int[] num = new int[N];

      StringTokenizer st = new StringTokenizer(br.readLine());
      for(int i = 0; i < N; i++){
        num[i] = Integer.parseInt(st.nextToken());
      }

      ArrayDeque<Integer> stack = new ArrayDeque<>();
      int[] answer = new int[N];
      for(int i = N - 1; i >= 0; i--)
      {
        while(!stack.isEmpty() && stack.peekLast() <= num[i]){
            stack.pollLast();
        }

        answer[i] = stack.isEmpty() ? -1 : stack.peekLast();
      stack.addLast(num[i]);
      }

      for(int i : answer){
        bw.write(i + " ");
      }

      bw.flush();
       
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int main() {

	int n = 0;
	scanf("%d", &n);

	vector<int> num(n, 0), answer(n, 0);
	for (int i = 0; i < n; i++) {
		scanf("%d", &num[i]);
	}

	stack<int> st;
	for (int i = n - 1; i >= 0; i--) {
		while (!st.empty() && st.top() <= num[i]) {
			st.pop();
		}
		answer[i] = st.empty() ? -1 : st.top();
		st.push(num[i]);
	}

	for (int& it : answer) {
		printf("%d ", it);
	}

	return 0;
}